### PR TITLE
Refactor merge block producer initialization for AuRaMerge

### PR DESCRIPTION
**Pull request by jmederosalvarado**
Makes changes discussed [here](https://github.com/NethermindEth/nethermind/pull/4256#discussion_r917364214)

## Changes:
- Refactor merge block producer initialization to allow for AuRa custom producer
- Don't process empty blocks on merge block producer if not in AuRa
- Throw if empty block processing fails in post-merge aura chain
- Refactor AuRaMerge block processing and producing to use BlockProducerEnvFactory, which now allows to use the same class of block processor to process suggested blocks and to produce blocks. As alarming as it sounds, previously, we used the standard `BlockProcessor` class for producing blocks and `AuRaMergeBlockProcessor` for validating and processing suggested blocks, this didn't cause any issues because both block processors behave in exactly the same way post-merge, however it is bad design, and there is currently one testing scenario where this is being a blocker.

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [x] Yes
- [ ] No

**In case you checked yes, did you write tests??**

pre-merge posdao tests: [NethermindEth/nethermind/actions/runs/3326835492](https://github.com/NethermindEth/nethermind/actions/runs/3326835492)
post-merge posdao tests: [NethermindEth/nethermind/actions/runs/3326836422](https://github.com/NethermindEth/nethermind/actions/runs/3326836422)

### CODE REVIEW

File: src/Nethermind/Nethermind.Merge.AuRa/AuRaPostMergeBlockProducer.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        public override Block PrepareEmptyBlock(BlockHeader parent, PayloadAttributes? payloadAttributes = null)
+        {
+            var block = base.PrepareEmptyBlock(parent, payloadAttributes);
+
+            // processing is only done to apply AuRa block rewards and rewrite posdao contracts
+            // TODO should we return null instead of empty block if we couldn't process?
+            if (TrySetState(parent.StateRoot))
+                block = ProcessPreparedBlock(block, null) ?? block;
+
+            return block;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Merge.AuRa/AuRaPostMergeBlockProducer.cs:50]: If processing fails we still return unprocessed block

**Review by LukaszRozmej — CHANGES_REQUESTED** [review]
Would make factory and producer by interface even if inheriting producer (wouldn't inherit factory)

File: src/Nethermind/Nethermind.Merge.AuRa/AuRaPostMergeBlockProducer.cs (hunk 1 above)
Comment (jmederosalvarado) [reply src/Nethermind/Nethermind.Merge.AuRa/AuRaPostMergeBlockProducer.cs:50]: You think it's better to return null? Maybe throw? I don't know how would it impact the GetPayloadHandler if we return null here, I'll check

**Comment by jmederosalvarado** [comment]
> Would make factory and producer by interface even if inheriting producer (wouldn't inherit factory)

I agree, I didn't do it like that to prevent even more additions and modifications to MergePlugin, I was trying to keep the changes as constrained to AuRaMergePlugin as possible.

File: src/Nethermind/Nethermind.Merge.AuRa/AuRaPostMergeBlockProducer.cs (hunk 1 above)
Comment (jmederosalvarado) [reply src/Nethermind/Nethermind.Merge.AuRa/AuRaPostMergeBlockProducer.cs]: Let's throw if this happens. It would be a bug.

**Comment by jmederosalvarado** [comment]
pre-merge posdao tests: https://github.com/NethermindEth/nethermind/actions/runs/3326835492
post-merge posdao tests: https://github.com/NethermindEth/nethermind/actions/runs/3326836422

File: src/Nethermind/Nethermind.Merge.AuRa/AuRaMergeBlockProducerEnvFactory.cs [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+                rewardCalculatorSource.Get(readOnlyTxProcessingEnv.TransactionProcessor),
+                TransactionsExecutorFactory.Create(readOnlyTxProcessingEnv),
+                readOnlyTxProcessingEnv.StateProvider,
+                readOnlyTxProcessingEnv.StorageProvider,
+                receiptStorage,
+                logManager,
+                _blockTree);
+        }
+
+        protected override TxPoolTxSource CreateTxPoolTxSource(
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Merge.AuRa/AuRaMergeBlockProducerEnvFactory.cs:88]: Seems redundant if we already kept code in one place thanks to below. While we are exposing some internal AuRa stuff to be used here it is fine.

```
return new StartBlockProducerAuRa(_auraApi!)
    .CreateStandardTxSourceForProducer(txProcessingEnv, constantContractsProcessingEnv);
```

File: src/Nethermind/Nethermind.Merge.AuRa/AuRaPostMergeBlockProducer.cs [hunk 3]
```diff
@@ -44,12 +44,10 @@ public override Block PrepareEmptyBlock(BlockHeader parent, PayloadAttributes? p
...
             var block = base.PrepareEmptyBlock(parent, payloadAttributes);
 
-            // processing is only done to apply AuRa block rewards and rewrite posdao contracts
-            // TODO should we return null instead of empty block if we couldn't process?
             if (TrySetState(parent.StateRoot))
-                block = ProcessPreparedBlock(block, null) ?? block;
+                return ProcessPreparedBlock(block, null) ?? throw new System.Exception("Couldn't process empty block");
 
-            return block;
+            throw new System.Exception("Couldn't process empty block");
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Merge.AuRa/AuRaPostMergeBlockProducer.cs:50]: Don't ever throw `System.Exception`, throw something meaningful

File: src/Nethermind/Nethermind.Merge.AuRa/AuRaMergeBlockProcessor.cs [hunk 4]
```diff
@@ -63,12 +60,10 @@ public class AuRaMergeBlockProcessor : AuRaBlockProcessor
...
             contractRewriter
         )
-    {
-        _poSSwitcher = poSSwitcher;
-    }
+    { }
 
     protected override TxReceipt[] ProcessBlock(Block block, IBlockTracer blockTracer, ProcessingOptions options) =>
-        _poSSwitcher.IsPostMerge(block.Header)
+        block.IsPostMerge
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Merge.AuRa/AuRaMergeBlockProcessor.cs:66]: This is probably enough here? @MarekM25

Comment (jmederosalvarado) [reply src/Nethermind/Nethermind.Merge.AuRa/AuRaMergeBlockProcessor.cs:66]: Check `MergePlugin.Init` method, there is a preprocessor steps that sets this property using the `PoSSwitcher`

File: src/Nethermind/Nethermind.Merge.AuRa/AuRaPostMergeBlockProducer.cs (hunk 3 above)
Comment (jmederosalvarado) [reply src/Nethermind/Nethermind.Merge.AuRa/AuRaPostMergeBlockProducer.cs:50]: Yeah, I was hoping that the exception type I'm looking for would already exist and you would point me to it

File: src/Nethermind/Nethermind.Merge.AuRa/AuRaMergeBlockProducerEnvFactory.cs (hunk 2 above)
Comment (jmederosalvarado) [reply src/Nethermind/Nethermind.Merge.AuRa/AuRaMergeBlockProducerEnvFactory.cs:88]: done

File: src/Nethermind/Nethermind.Merge.AuRa/AuRaMergeBlockProcessor.cs (hunk 4 above)
Comment (jmederosalvarado) [reply src/Nethermind/Nethermind.Merge.AuRa/AuRaMergeBlockProcessor.cs:66]: This change is already running in a shadowfork, and it is working just fine

**Comment by jmederosalvarado** [comment]
posdao tests after updating with master:
https://github.com/NethermindEth/nethermind/actions/runs/3768637286
https://github.com/NethermindEth/nethermind/actions/runs/3768636480

**Review by LukaszRozmej — APPROVED** [review]

**Pull request merged**